versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java

```diff
@@ -0,0 +1,416 @@
+import com.mongodb.ConnectionString;
+import com.mongodb.ErrorCategory;
+import com.mongodb.MongoClientSettings;
+import com.mongodb.MongoWriteException;
+import com.mongodb.client.FindIterable;
+import com.mongodb.client.MongoClient;
+import com.mongodb.client.MongoClients;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.MongoDatabase;
+import com.mongodb.client.model.Filters;
+import com.mongodb.client.model.UpdateOptions;
+import com.mongodb.client.result.DeleteResult;
+import com.mongodb.client.result.InsertManyResult;
+import com.mongodb.client.result.InsertOneResult;
+import com.mongodb.client.result.UpdateResult;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.bson.Document;
+import org.bson.types.Binary;
+import org.projectnessie.versioned.Hash;
+import org.projectnessie.versioned.ReferenceConflictException;
+import org.projectnessie.versioned.persist.adapter.CommitLogEntry;
+import org.projectnessie.versioned.persist.adapter.KeyList;
+import org.projectnessie.versioned.persist.adapter.KeyListEntity;
+import org.projectnessie.versioned.persist.adapter.KeyWithType;
+import org.projectnessie.versioned.persist.adapter.spi.DatabaseAdapterUtil;
+import org.projectnessie.versioned.persist.nontx.NonTransactionalDatabaseAdapter;
+import org.projectnessie.versioned.persist.nontx.NonTransactionalOperationContext;
+import org.projectnessie.versioned.persist.serialize.AdapterTypes.GlobalStateLogEntry;
+import org.projectnessie.versioned.persist.serialize.AdapterTypes.GlobalStatePointer;
+import org.projectnessie.versioned.persist.serialize.ProtoSerialization;
+
+public class MongoDatabaseAdapter
+    extends NonTransactionalDatabaseAdapter<MongoDatabaseAdapterConfig> {
+
+  private static final String GLOBAL_POINTER = "global_pointer";
+  private static final String GLOBAL_LOG = "global_log";
+  private static final String COMMIT_LOG = "commit_log";
+  private static final String KEY_LIST = "key_list";
+
+  private final ByteString keyPrefix;
+  private final byte[] globalPointerKey;
+
+  private final MongoClient mongoClient;
+  private final MongoCollection<Document> globalPointers;
+  private final MongoCollection<Document> globalLog;
+  private final MongoCollection<Document> commitLog;
+  private final MongoCollection<Document> keyLists;
+
+  protected MongoDatabaseAdapter(MongoDatabaseAdapterConfig config) {
+    super(config);
+
+    keyPrefix = ByteString.copyFromUtf8(config.getKeyPrefix());
+    globalPointerKey = keyPrefix.toByteArray();
+
+    ConnectionString cs = new ConnectionString(config.getConnectionString());
+    MongoClientSettings settings = MongoClientSettings.builder().applyConnectionString(cs).build();
+
+    mongoClient = MongoClients.create(settings);
```

Comment:
`MongoDatabaseAdapterConfig` needs a functionality to provide a `MongoClient` instance to `MongoDatabaseAdapterConfig` (like `RocksDatabaseAdapter` does). I.e. if the `MongoDatabaseAdapterConfig` provides a non-null one, use that, otherwise create one. Probably better to wrap that `MongoClient` in a class similar to `RocksDbInstance`, which would also care about inquiring the collections (the `database.getCollection()` calls / keep the `MongoCollection` instances.

In other words: this constructor should be fast and cheap, if the MongoClient is provided via the config.

Reply:
I see. Did not realize this was the intention :)

However, why would we pass proper "configuration" settings to this class in this case? Why not always pass the "client" / "instance" object and let the factory configure it?

Reply:
I'm thinking about splitting "config" and "database instance" method at the factoring level :hourglass:

Reply:
I took an approach similar to `RocksDbInstance` in cc90e03fe60a625171481980ca86e887f73a7abe for now.

Still thinking about refactoring the config/factoring classes :thinking: